src/algebra/magma_algebra.lean

```diff
@@ -0,0 +1,189 @@
+  ..(infer_instance : mul_zero_class _), }
+
+lemma smul_mul_assoc (t : R) (a b : magma_algebra R M) : (t • a) * b = t • (a * b) :=
+begin
+  ext,
+  simp only [mul_def, mul_sum, smul_apply, sum_smul_index, implies_true_iff, eq_self_iff_true,
+    single_zero, sum_apply, smul_eq_mul, single_apply, mul_ite, mul_zero, sum_zero, zero_mul,
+    mul_assoc],
+end
+
+/-- If the coefficients are commutative amongst themselves, they also commute with the
+multiplication. -/
+lemma mul_smul_comm {R : Type u} [comm_semiring R] (t : R) (a b : magma_algebra R M) :
+  a * (t • b) = t • (a * b) :=
+begin
+  ext,
+  simp only [mul_def, mul_sum, smul_apply, sum_smul_index, implies_true_iff, eq_self_iff_true,
+    single_zero, sum_apply, smul_eq_mul, single_apply, mul_ite, mul_zero, mul_left_comm],
+end
+
+lemma mul_sum_single
+  {A : Type w} [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A]
+  (m : M) (t : R) {b : magma_algebra R M} {f : mul_hom M A} :
+  ((single m t) * b).sum (λ m t, t • f m) =
+  ((single m t).sum (λ m t, t • f m)) * (b.sum (λ m t, t • f m)) :=
+begin
+  simp only [mul_def, zero_mul, sum_single_index, sum_zero, single_zero, zero_smul],
+  apply induction_linear b,
+  { simp only [sum_zero_index, mul_zero], },
+  { intros b₁ b₂ hb₁ hb₂,
+    simp only [sum_add_index, hb₁, hb₂, left_distrib, add_smul, single_add, implies_true_iff,
+      eq_self_iff_true, zero_smul, forall_3_true_iff, single_eq_zero, mul_zero], },
+  { intros m' t',
+    simp only [mul_def, sum_single_index, single_eq_zero, zero_smul, zero_mul, mul_zero],
+    rw [mul_hom.map_mul, non_unital_non_assoc_algebra.smul_mul_assoc,
+      non_unital_non_assoc_algebra.mul_smul_comm, mul_smul], },
+end
+
+lemma mul_sum {A : Type w} [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A]
+  {a b : magma_algebra R M} {f : mul_hom M A} :
+  (a * b).sum (λ m t, t • f m) = (a.sum (λ m t, t • f m)) * (b.sum (λ m t, t • f m)) :=
+begin
+  apply induction_linear a,
+  { simp only [zero_mul, sum_zero_index], },
+  { intros a₁ a₂ ha₁ ha₂,
+    simp only [mul_def, sum_add_index, add_smul, implies_true_iff, eq_self_iff_true, zero_smul,
+      forall_3_true_iff],
+    rw sum_add_index,
+    { rw sum_add_index,
+      { simp only [mul_def] at ha₁ ha₂, simp only [ha₁, ha₂, right_distrib], },
+      { simp only [implies_true_iff, eq_self_iff_true, zero_smul], },
+      { simp only [add_smul, eq_self_iff_true, forall_3_true_iff], }, },
+    { simp only [zero_mul, implies_true_iff, eq_self_iff_true, sum_zero, single_zero], },
+    { intros m t t', simp only [add_mul, single_add, sum_add], }, },
+  { intros m t, apply mul_sum_single, },
+end
+
+instance {R : Type u} [comm_semiring R] : non_unital_non_assoc_algebra R (magma_algebra R M) :=
+{ smul_mul_assoc' := smul_mul_assoc,
+  mul_smul_comm'  := mul_smul_comm,
+  ..(infer_instance : semimodule R _) }
+
+variables (R)
+
+/-- The embedding of a magma into its magma algebra. -/
+def of : mul_hom M (magma_algebra R M) :=
```

Comment:
I guess this statement and `lift` below are currently not stated for `monoid_algebra`, but we could probably add them.

Reply:
I think it is worth adding `of` and `lift` here to `monoid_algebra`. I claim this justifies the creation of:
```lean
set_option old_structure_cmd true

structure linear_mul_hom (R : Type u) (A : Type v) (B : Type w)
  [semiring R] [non_unital_non_assoc_semiring A] [non_unital_non_assoc_semiring B]
  [module R A] [module R B] extends A →ₗ[R] B, mul_hom A B
```
By using `module` instead of `algebra` in this definition, we avoid the `algebra` issue and lose nothing really.

Thoughts?

Reply:
@riccardobrasca actually could have done with an object of this type very recently, for `linear_map.fst` and `linear_map.snd`.

I think `non_unital_alg_hom` might be a better name for that structure, and it suggests we should probably have `non_unital_ring_hom` too.

Reply:
Adding a new bundled hom comes with _lots_ of boilerplate unfortunately.

Reply:
I'll give it a go!

Reply:
Oh, yes, I proved needed exactly this for `linear_map.inl` and `linear_map.snd`!